src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs

```diff
@@ -167,43 +157,149 @@ internal int EndRead(IAsyncResult asyncResult)
 
         internal IAsyncResult BeginWrite(byte[] buffer, int offset, int count, AsyncCallback asyncCallback, object asyncState)
         {
-            var lazyResult = new LazyAsyncResult(this, asyncState, asyncCallback);
-            ProcessWrite(buffer, offset, count, lazyResult);
-            return lazyResult;
+            return TaskToApm.Begin(WriteAsync(buffer, offset, count, CancellationToken.None), asyncCallback, asyncState);
         }
 
-        internal void EndWrite(IAsyncResult asyncResult)
+        internal void EndWrite(IAsyncResult asyncResult) => TaskToApm.End(asyncResult);
+
+        internal Task WriteAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
         {
-            if (asyncResult == null)
+            ValidateParameters(buffer, offset, count);
+
+            SslWriteAsync writeAdapter = new SslWriteAsync(_sslState, cancellationToken);
+
+            return WriteAsyncInternal(writeAdapter, buffer, offset, count);
+        }
+
+        private Task WriteAsyncInternal<T>(T writeAdapter, byte[] buffer, int offset, int count)
+            where T : struct, ISslWriteAdapter
+        {
+            _sslState.CheckThrow(authSuccessCheck: true, shutdownCheck: true);
+
+            if (count == 0 && !SslStreamPal.CanEncryptEmptyMessage)
             {
-                throw new ArgumentNullException(nameof(asyncResult));
+                // If it's an empty message and the PAL doesn't support that, we're done.
+                return Task.CompletedTask;
             }
 
-            LazyAsyncResult lazyResult = asyncResult as LazyAsyncResult;
-            if (lazyResult == null)
+            if (Interlocked.Exchange(ref _nestedWrite, 1) == 1)
             {
-                throw new ArgumentException(SR.Format(SR.net_io_async_result, asyncResult.GetType().FullName), nameof(asyncResult));
+                throw new NotSupportedException(SR.Format(SR.net_io_invalidnestedcall, "WriteAsync", "write"));
             }
 
-            if (Interlocked.Exchange(ref _nestedWrite, 0) == 0)
+            if (count <= _sslState.MaxDataSize)
             {
-                throw new InvalidOperationException(SR.Format(SR.net_io_invalidendcall, "EndWrite"));
+                //Single write
+                Task t = WriteSingleChunk(writeAdapter, buffer, offset, count);
```

Comment:
It looks like WriteSingleChunk can throw synchronously.  Don't we need to handle that here?

Reply:
Yeah, I think the best way is to wrap that throw in inside the WriteSingleChunk into a

Task.FromException and return that, then it will be picked up by the normal flow without more try catches?

Reply:
Fixed by returning the exception inside a task